devito/profiler.py

```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
+                    brackets += 1
+                    idx += 1
+                elif char is ' ' and brackets == 0 and len(cur_load) > 0:
+                    loads[cur_load] = True
+                    cur_load = ""
+                    idx += 1
+                else:
+                    cur_load += char
+                    idx += 1
+
+        if len(cur_load) > 0:
+            loads[cur_load] = True
+
+        return flops
+
+    def get_class(self, choice):
         """Returns a :class:`ctypes.Structure` subclass defining our structure
+        for Ois or Timings
 
-        :returns: A :class:`Timings` class definition
+        :param choice: Profiler.OIS or Profiler.TIMINGS
+
+        :returns: A class definition
         """
-        return type("Timings", (Structure,), {"_fields_": self.fields})
+        if choice == Profiler.TIME:
+            name = "Timings"
+            fields = self.t_fields
+        elif choice == Profiler.FLOP:
+            name = "Flops"
+            fields = self.f_fields
+        else:
+            error("Wrong choice")
 
-    @property
-    def as_cgen_struct(self):
+        return type(name, (Structure,), {"_fields_": fields})
+
+    def as_cgen_struct(self, choice):
         """Returns the :class:`cgen.Struct` relative to the profiler
 
         :returns: The struct
         """
         fields = []
+        s_name = None
 
-        for name, _ in self.fields:
-            fields.append(Value("double", name))
+        if choice == Profiler.TIME:
+            s_name = "profiler"
+            for name, _ in self.t_fields:
+                fields.append(Value("double", name))
+        elif choice == Profiler.FLOP:
+            s_name = "flops"
+            for name, _ in self.f_fields:
+                fields.append(Value("long long", name))
+        else:
+            error("Wrong choice")
 
-        return Struct("profiler", fields)
+        return Struct(s_name, fields)
 
-    @property
-    def as_ctypes_pointer(self):
-        """Returns a pointer to the ctypes structure
+    def as_ctypes_pointer(self, choice):
+        """Returns a pointer to the ctypes structure for the chosen
+        metric
+
+        :param choice: The structure needed
 
         :returns: The pointer
         """
-        self.timings = self.get_timings_class()
+        struct = self.get_class(choice)()
 
-        return byref(self.timings)
+        if choice == Profiler.TIME:
+            self._timings = struct
+        elif choice == Profiler.FLOP:
+            self._flops = struct
+
+        return byref(struct)
 
     @property
-    def as_dictionary(self):
+    def timings(self):
         """Returns the timings as a python dictionary
 
         :returns: A dictionary containing the timings
         """
-        if not self.timings:
+        if not self._timings:
+            return {}
+
+        return dict((field, getattr(self._timings, field))
+                    for field, _ in self._timings._fields_)
```

Comment:
`self._timings` is a ctypes struct. `self._timings._fields_` contains a list of tuples `(name, type)`. I get the name from there and the value with `getattr`.

It's simply a conversion from a ctypes struct to a `dict`